helsingborg-stad/app-mitt-helsingborg
# Feat/form values context

## Explain the changes you’ve made

Add a context to our Form container, so that any component in a Form can get access to some functions and values that 'belong' to the form; such as the current answers, current form position, and the formNavigation actions. 

## Explain why these changes are made

This context is added to reduce the amount of prop drilling; currently we pass a lot of props through the Step component, and then further down through the Footer and FormField components and so on. Using this context, a lot of this can be removed and instead gotten by the useContext hook. The main motivation is the SummaryList, which needs access to all the answers and questions in order to summarize them, which today necessitates a lot of prop drilling, which all the other input components have no use of. I thought of this while implementing the smart keyboard logic; for the SummaryList especially, this becomes much easier given a context like this one.

Wrapping a Form in a context like this is not my original idea, but comes from the [Formik library](https://github.com/formium/formik), which is used for the formbuilder. While working on the that, I eventually noticed this, and could use it to reduce the prop drilling by a lot. This PR only introduces the context, no refactoring to remove props is done. 

## Explain your solution

In a separate file, I define a context called FormValueContext (not really happy with this name, any suggestions are very welcome... we already have a FormContext, which would be most natural), and define its types quite explicitly. This is then used in the Form container, wrapping all its other children (i.e. all the steps), and passed the values/functions from the useForm hook.  

Then, in any place inside the Form, one can get access to these by using the context, const { formState, formNavigation } = useContext(FormValueContext), as usual. 

## How to test the changes?

Since this just adds a wrapping context, there is no direct way of testing it "out of the box", so to speak. One thing one can do is to go into for example the Step component, and add some lines that uses the context, and perhaps logs the values to check that indeed the correct formState is retrieved. I tested it like that, but of course that should not be a part of the PR. 

## Was this feature tested in the following environments?
- [] Storybook on a iOS device/simulator.
- [] Building the Application on a iOS device/simulator.
